src/window.c

```diff
@@ -4416,6 +4418,13 @@ close_others(
     for (wp = firstwin; win_valid(wp); wp = nextwp)
     {
 	nextwp = wp->w_next;
+
+	// autocommands messed this one up
+	if (old_curwin != curwin && win_valid(old_curwin))
+	    curwin = old_curwin;
+	if (old_curbuf != curwin->w_buffer)
+	    curbuf = curwin->w_buffer;
```

Comment:
Keeping the `old_curbuf != curwin->w_buffer` condition doesn't look correct:
- If the previous condition succeeds, `curwin` is now equal to `old_curwin`, and `old_curbuf` is already equal to `curwin->w_buffer` unless some other autocommand changed `old_curwin`'s buffer,  but `curbuf` is still equal to the previous `curwin->w_buffer`, and not necessarily equal to the current `curwin->w_buffer`.
- If the previous condition fails, `curwin` is unchanged, and `curbuf` should still be equal to `curwin->w_buffer` as it was before.

```suggestion
	{
	    curwin = old_curwin;
	    curbuf = curwin->w_buffer;
	}
```